temporal-sdk/src/main/java/io/temporal/internal/sync/DeterministicRunnerImpl.java

```diff
@@ -343,6 +372,9 @@ public void close() {
         throw new Error("Unexpected failure stopping coroutine", e);
       }
     }
+
+    closeFuture.complete(null);
+    closed = true;
```

Comment:
Javadoc says this is to be used under lock, but this is section of code is not locked (can make synchronized or use `AtomicBoolean` if it's important to be unlocked while waiting on `threadFutures`)

Also, some languages have compile-time guards you can use with annotations that, when you say a mutex/lock applies to a class member, it statically analyzes that only that class member is used under that lock. Does Java have anything like that?